FlowCrypt/src/main/java/com/flowcrypt/email/database/dao/RecipientDao.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
+ * Contributors: DenBond7
+ */
+
+package com.flowcrypt.email.database.dao
+
+import android.database.Cursor
+import androidx.lifecycle.LiveData
+import androidx.room.Dao
+import androidx.room.Query
+import androidx.room.Transaction
+import com.flowcrypt.email.database.entity.RecipientEntity
+import com.flowcrypt.email.database.entity.relation.RecipientWithPubKeys
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * This object describes a logic of work with [RecipientEntity].
+ *
+ * @author DenBond7
+ * Date: 17.05.2017
+ * Time: 12:22
+ * E-mail: DenBond7@gmail.com
+ */
+@Dao
+interface RecipientDao : BaseDao<RecipientEntity> {
+  @Query("SELECT * FROM recipients")
+  suspend fun getAllRecipients(): List<RecipientEntity>
+
+  @Query("SELECT * FROM recipients")
+  fun getAllRecipientsLD(): LiveData<List<RecipientEntity>>
+
+  @Query("SELECT recipients.* FROM recipients INNER JOIN public_keys ON recipients.email = public_keys.recipient GROUP BY recipients.email ORDER BY recipients._id")
+  fun getAllRecipientsWithPgpFlow(): Flow<List<RecipientEntity>>
+
+  @Query("SELECT recipients.* FROM recipients INNER JOIN public_keys ON recipients.email = public_keys.recipient GROUP BY recipients.email ORDER BY recipients._id")
+  suspend fun getAllRecipientsWithPgp(): List<RecipientEntity>
+
+  @Query("SELECT recipients.* FROM recipients INNER JOIN public_keys ON recipients.email = public_keys.recipient WHERE (email LIKE :searchPattern OR name LIKE :searchPattern) GROUP BY recipients.email ORDER BY recipients._id")
+  suspend fun getAllRecipientsWithPgpWhichMatched(searchPattern: String): List<RecipientEntity>
+
+  @Query("SELECT * FROM recipients WHERE email = :email")
+  suspend fun getRecipientByEmailSuspend(email: String): RecipientEntity?
+
+  @Query("SELECT * FROM recipients WHERE email = :email")
+  fun getRecipientByEmail(email: String): RecipientEntity?
+
+  @Transaction
+  @Query("SELECT * FROM recipients WHERE email = :email")
+  fun getRecipientsWithPubKeysByEmailsLD(email: String): LiveData<RecipientWithPubKeys?>
+
+  @Transaction
+  @Query("SELECT * FROM recipients WHERE email IN (:emails)")
+  fun getRecipientsWithPubKeysByEmails(emails: Collection<String>): List<RecipientWithPubKeys>
+
+  @Query("SELECT * FROM recipients WHERE email LIKE :searchPattern ORDER BY last_use DESC")
+  fun getFilteredCursor(searchPattern: String): Cursor?
+
+  @Query("DELETE FROM recipients")
+  suspend fun deleteAll(): Int
+
+  @Transaction
+  @Query("SELECT * FROM recipients WHERE email = :email")
+  suspend fun getRecipientWithPubKeysByEmailSuspend(email: String): RecipientWithPubKeys?
```

Comment:
> Why some queries are @Transaction and some not?
> What is benefit of @Transaction for SELECT?

`RecipientWithPubKeys` it is a object that [define one-to-many relationships](https://developer.android.com/training/data-storage/room/relationships#one-to-many). 

Querying such an object requires Room to run two queries, so we should add the `@Transaction` annotation to this method to ensure that the whole operation is performed atomically.

Reply:
That's confusing. Because the query `SELECT * FROM recipients WHERE email = :email` would, in SQL, mean `get all records that match email from this one table, and don't touch the pubkeys table`. But this returns `RecipientWithPubKeys`. Is there some magic behind that internally does some sort of second query on the pubkeys table and matches it up?